ros2snap

```diff
@@ -1,112 +1,345 @@
+export PYTHONPATH=$mydir/install/lib/python2.7/dist-packages:$PYTHONPATH
+export PKG_CONFIG_PATH=$mydir/install/lib/pkgconfig:$mydir/install/lib/x86_64-linux-gnu/pkgconfig:PKG_CONFIG_PATH
+export CMAKE_PREFIX_PATH=$mydir/install:$CMAKE_PREFIX_PATH
+"""
+      #setup += "export _CATKIN_SETUP_DIR=%s\n" % local_setup_dir[:len(local_setup_dir)-1]
+      #setup += ". %ssetup.sh\n" % local_setup_dir
+
+    self.environment_script = """#!/bin/bash
+mydir=$(dirname $(dirname $(builtin cd "`dirname "${BASH_SOURCE[0]}"`" > /dev/null && pwd)))
+%s
+export ROS_MASTER_URI=http://localhost:11311
+export LD_LIBRARY_PATH=$mydir/usr/lib/x86_64-linux-gnu:$mydir/usr/lib:$LD_LIBRARY_PATH
+export PATH=$mydir/usr/bin:$PATH
+export PYTHONPATH=$mydir/usr/lib/python2.7/dist-packages:$PYTHONPATH
+export PKG_CONFIG_PATH=$mydir/usr/lib/pkgconfig:$mydir/usr/lib/x86_64-linux-gnu/pkgconfig:$PKG_CONFIG_PATH
+""" % setup
+
+    self.cache = apt.Cache()
+
+  def resolve_and_copy(self, key):
+    # Resolve key to system package name
+    # TODO: Cross-platform solution for args
+    key_entry = bloom.generators.common.resolve_rosdep_key(key, "ubuntu", "trusty")
+    if key_entry is None or (key_entry[0] is None) or (len(key_entry[0]) == 0):
+      return
+
+    apt_name = key_entry[0][0]
+    self.copy_from_apt_cache(apt_name)
+
+  def copy_files(self, run_dep_files):
+    # Copy all the files
+    for dep_path in run_dep_files:
+      if os.path.isfile(dep_path):
+        fullpath = self.snappy_root + dep_path
+        check_create_dir(os.path.dirname(fullpath))
+        shutil.copy2(dep_path, fullpath)
+
+  def copy_from_apt_cache(self, apt_name):
+    # Use apt python API to get all files
+    run_pkg = self.cache[apt_name]
+
+    # TODO for performance: exclusion rule for top-level packages (e.g. python)
+    # that snappy instance already has
+    if not apt_name.startswith("ros-" + self.distro + "-"):
+      versions = run_pkg.versions
+      # system packages
+      if len(versions) > 0:
+        version = versions[0]
+        for dependency in version.dependencies:
+          key = dependency[0].name
+          if key in self.copied_packages or not self.cache.has_key(key):
+            continue
+          self.copied_packages.add(key)
+          self.copy_from_apt_cache(key)
+
+    # TODO: Catch more errors
+    # TODO: install missing run deps with package manager
+    # TODO: What about local installs?
+
+    self.copy_files(run_pkg.installed_files)
+
+  def copy_recursive_dependencies(self, package):
+    run_dep_keys = [dep.name for dep in package.run_depends]
+
+    for run_dep in run_dep_keys:
+      if run_dep in self.copied_packages:
+        continue
+      self.resolve_and_copy(run_dep)
+      self.copied_packages.add(run_dep)
+      # Get the package.xml of the run dependencies if it's a ROS package
+      ros_path = os.path.dirname(os.getenv("ROS_ROOT")) + "/" + run_dep
+      packages = catkin_pkg.packages.find_packages(ros_path)
```

Comment:
If I understand it correctly `ros_path` should point to `<prefix>/share/<run_dep>` and will only contain a single package. Therefore instead of trying to crawl for multiple packages you could call `catkin_pkg.package.parse_package(ros_path)` instead and get a single package back.